src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -271,23 +271,219 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
+
+        if (actualSize.x > targetSize.x)
+        {
+            scalingRequired = true;
+            offset.x = (overhang.left - overhang.right) * 0.5f;
+            scale.x = targetSize.x / actualSize.x;
+            scale.y = scale.x;
+        }
+        if (actualSize.y > targetSize.y)
+        {
+            scalingRequired = true;
+            offset.y = (overhang.top - overhang.bottom) * 0.5f;
+            scale.x = std::min(scale.x, targetSize.y / actualSize.y);
+            scale.y = scale.x;
+        }
+
+        offset.x *= scale.x;
+        offset.y *= scale.y;
+
+        // Due to D2D1_DRAW_TEXT_OPTIONS_NO_SNAP our baseline won't get
+        // pixel aligned anymore and so we have to do it ourselves here.
+        offset.x = roundf(offset.x * _r.pixelPerDIP) * _r.dipPerPixel;
+        offset.y = roundf(offset.y * _r.pixelPerDIP) * _r.dipPerPixel;
+    }
+
+    // !!! IMPORTANT !!!
+    // DirectWrite/2D snaps the baseline to whole pixels, which is something we technically
+    // want (it makes text look crisp), but fails in weird ways if `scalingRequired` is true.
+    // As our scaling matrix's dx/dy (center point) is based on the `origin` coordinates
+    // each cell we draw gets a unique, fractional baseline which gets rounded differently.
+    // I'm not 100% sure why that happens, since `origin` is always in full pixels...
+    // But this causes wide glyphs to draw as tiles that are potentially misaligned vertically by a pixel.
+    // The resulting text rendering looks especially bad for ligatures like "====" in Cascadia Code,
+    // where every single "=" might be blatantly misaligned vertically (same for any box drawings).
+    WI_SetFlagIf(options, D2D1_DRAW_TEXT_OPTIONS_NO_SNAP, scalingRequired);
+
+    const f32x2 inverseScale{ 1.0f - scale.x, 1.0f - scale.y };
+    const f32x2 halfSize{ layoutBox.x * 0.5f, layoutBox.y * 0.5f };
+
+    for (u32 i = 0; i < cellCount; ++i)
     {
         const auto coord = coords[i];
 
         D2D1_RECT_F rect;
-        rect.left = static_cast<float>(coord.x) * static_cast<float>(USER_DEFAULT_SCREEN_DPI) / static_cast<float>(_r.dpi);
-        rect.top = static_cast<float>(coord.y) * static_cast<float>(USER_DEFAULT_SCREEN_DPI) / static_cast<float>(_r.dpi);
+        rect.left = static_cast<float>(coord.x) * _r.dipPerPixel;
+        rect.top = static_cast<float>(coord.y) * _r.dipPerPixel;
         rect.right = rect.left + _r.cellSizeDIP.x;
         rect.bottom = rect.top + _r.cellSizeDIP.y;
 
         D2D1_POINT_2F origin;
-        origin.x = rect.left - i * _r.cellSizeDIP.x;
-        origin.y = rect.top;
+        origin.x = rect.left - i * _r.cellSizeDIP.x + offset.x;
+        origin.y = rect.top + offset.y;
 
         _r.d2dRenderTarget->PushAxisAlignedClip(&rect, D2D1_ANTIALIAS_MODE_ALIASED);
         _r.d2dRenderTarget->Clear();
+        if (scalingRequired)
+        {
+            // TODO: dx/dy should technically be pixel snapped

```

Comment:
Is this TODO now or TODO later?

Reply:
I wrote that comment for future-someone and not specifically for us.
I'll try to figure out the correct approach for this before it gets merged, but if I don't it's not that important anyways - the scaled glyphs are pretty crisp either way.